File: py-core/node/message.ts
import { Address, NodeId, TunnelPortId } from "@core/net";
import {
    EmptySerdeable,
    EnumSerdeable,
    ObjectSerdeable,
    RemainingBytesSerdeable,
    TransformSerdeable,
    Uint32Serdeable,
    Utf8Serdeable,
    VariantSerdeable,
} from "@core/serde";

export class SendHello {
    index = 1;
    address: Address;

    constructor(args: { address: Address }) {
        this.address = args.address;
    }

    static readonly serdeable = new TransformSerdeable(
        new ObjectSerdeable({ address: Address.serdeable }),
        (obj) => new SendHello(obj),
        (obj) => obj,
    );
}

export class SocketDescriptor {
    descriptor: number;

    constructor(args: { descriptor: number }) {
        this.descriptor = args.descriptor;
    }

    static readonly serdeable = new TransformSerdeable(
        new ObjectSerdeable({ descriptor: new Uint32Serdeable() }),
        (obj) => new SocketDescriptor(obj),
        (obj) => obj,
    );
}

export enum SocketProtocol {
    TUNNEL = 1,
    TRUSTED = 2,
    STREAM = 3,
}
const SocketProtocolSerdeable = new EnumSerdeable<SocketProtocol>(SocketProtocol);

export class OpenSocket {
    index = 2;
    protocol: SocketProtocol;
    port: TunnelPortId;

    constructor(args: { protocol: SocketProtocol; port: TunnelPortId }) {
        this.protocol = args.protocol;
        this.port = args.port;
    }

    static readonly serdeable = new TransformSerdeable(
        new ObjectSerdeable({ protocol: SocketProtocolSerdeable, port: TunnelPortId.serdeable }),
        (obj) => new OpenSocket(obj),
        (obj) => obj,
    );
}

export class ConnectSocket {
    index = 3;
    protocol: SocketProtocol;
    remote: NodeId;
    port: TunnelPortId;

    constructor(args: { protocol: SocketProtocol; remote: NodeId; port: TunnelPortId }) {
        this.protocol = args.protocol;
        this.remote = args.remote;
        this.port = args.port;
    }

    static readonly serdeable = new TransformSerdeable(
        new ObjectSerdeable({
            protocol: SocketProtocolSerdeable,
            remote: NodeId.serdeable,
            port: TunnelPortId.serdeable,
        }),
        (obj) => new ConnectSocket(obj),
        (obj) => obj,
    );
}

export class CloseSocket {
    index = 4;
    descriptor: SocketDescriptor;

    constructor(args: { descriptor: SocketDescriptor }) {
        this.descriptor = args.descriptor;
    }

    static readonly serdeable = new TransformSerdeable(
        new ObjectSerdeable({ descriptor: SocketDescriptor.serdeable }),
        (obj) => new CloseSocket(obj),
        (obj) => obj,
    );
}

export class SendData {
    index = 5;
    descriptor: SocketDescriptor;
    payload: Uint8Array;

    constructor(args: { descriptor: SocketDescriptor; payload: Uint8Array }) {
        this.descriptor = args.descriptor;
        this.payload = args.payload;
    }

    static readonly serdeable = new TransformSerdeable(
        new ObjectSerdeable({
            descriptor: SocketDescriptor.serdeable,
            payload: new RemainingBytesSerdeable(),
        }),
        (obj) => new SendData(obj),
        (obj) => obj,
    );
}

export class Terminate {
    index = 6;
    static readonly serdeable = new TransformSerdeable(
        new EmptySerdeable(),
        () => new Terminate(),
        () => ({}),
    );
}

export type RequestBodyType = SendHello | OpenSocket | ConnectSocket | CloseSocket | SendData | Terminate;
export class RequestBody {
    index = 1;
    body: RequestBodyType;

    constructor(args: { body: RequestBodyType }) {
        this.body = args.body;
    }

    static readonly serdeable = new VariantSerdeable(
        [
            SendHello.serdeable,
            OpenSocket.serdeable,
            ConnectSocket.serdeable,
            CloseSocket.serdeable,
            SendData.serdeable,
            Terminate.serdeable,
        ],
        (obj) => obj.index,
    );
}

export class OperationSuccess {
    index = 1;

    static readonly serdeable = new TransformSerdeable(
        new EmptySerdeable(),
        () => new OperationSuccess(),
        () => ({}),
    );
}

export class OperationFailure {
    index = 2;
    trace: string;

    constructor(args: { trace: string }) {
        this.trace = args.trace;
    }

    static readonly serdeable = new TransformSerdeable(
        new ObjectSerdeable({ trace: new Utf8Serdeable(new Uint32Serdeable()) }),
        (obj) => new OperationFailure(obj),
        (obj) => obj,
    );
}

export class SocketOpened {
    index = 3;
    descriptor: SocketDescriptor;

    constructor(args: { descriptor: SocketDescriptor }) {
        this.descriptor = args.descriptor;
    }

    static readonly serdeable = new TransformSerdeable(
        new ObjectSerdeable({ descriptor: SocketDescriptor.serdeable }),
        (obj) => new SocketOpened(obj),
        (obj) => obj,
    );
}

export class SocketReceived {
    index = 4;
    descriptor: SocketDescriptor;
    payload: Uint8Array;

    constructor(args: { descriptor: SocketDescriptor; payload: Uint8Array }) {
        this.descriptor = args.descriptor;
        this.payload = args.payload;
    }

    static readonly serdeable = new TransformSerdeable(
        new ObjectSerdeable({
            descriptor: SocketDescriptor.serdeable,
            payload: new RemainingBytesSerdeable(),
        }),
        (obj) => new SocketReceived(obj),
        (obj) => obj,
    );
}

export class SocketClosed {
    index = 5;
    descriptor: SocketDescriptor;

    constructor(args: { descriptor: SocketDescriptor }) {
        this.descriptor = args.descriptor;
    }

    static readonly serdeable = new TransformSerdeable(
        new ObjectSerdeable({ descriptor: SocketDescriptor.serdeable }),
        (obj) => new SocketClosed(obj),
        (obj) => obj,
    );
}

export type ResponseBodyType = OperationSuccess | OperationFailure | SocketOpened | SocketReceived | SocketClosed;

export class ResponseBody {
    index = 2;
    body: ResponseBodyType;

    constructor(args: { body: ResponseBodyType }) {
        this.body = args.body;
    }

    static readonly serdeable = new VariantSerdeable(
        [
            OperationSuccess.serdeable,
            OperationFailure.serdeable,
            SocketOpened.serdeable,
            SocketReceived.serdeable,
            SocketClosed.serdeable,
        ],
        (obj) => obj.index,
    );
}

export class MessageDescriptor {
    descriptor: number;

    constructor(args: { descriptor: number }) {
        this.descriptor = args.descriptor;
    }

    static readonly serdeable = new TransformSerdeable(
        new ObjectSerdeable({ descriptor: new Uint32Serdeable() }),
        (obj) => new MessageDescriptor(obj),
        (obj) => obj,
    );
}

export class Message {
    descriptor: MessageDescriptor;
    body: RequestBody | ResponseBody;

    constructor(args: { descriptor: MessageDescriptor; body: RequestBody | ResponseBody }) {
        this.descriptor = args.descriptor;
        this.body = args.body;
    }

    static readonly serdeable = new ObjectSerdeable({
        descriptor: MessageDescriptor.serdeable,
        body: new VariantSerdeable([RequestBody.serdeable, ResponseBody.serdeable], (obj) => obj.index),
    });
}

File: py-core/py_core/ipc/message.py
from dataclasses import dataclass
from enum import Enum

from py_core.net import Address, Cost, NodeId, TunnelPort
from py_core.serde import UInt32, Utf8, Variant, DeriveSerde, RemainingBytes
from py_core.serde.enum import EnumSerde


@dataclass(frozen=True)
class MessageDescriptor(DeriveSerde):
    id: UInt32


@dataclass
class SendHello(DeriveSerde):
    """
    Response: OperationSuccess | OperationFailure
    """

    descriptor: MessageDescriptor
    address: Address
    link_cost: Cost


@dataclass(frozen=True)
class ServerDescriptor(DeriveSerde):
    id: UInt32


@dataclass(frozen=True)
class SocketDescriptor(DeriveSerde):
    id: UInt32


class SocketProtocolValue(Enum):
    TUNNEL = 1
    TRUSTED = 2
    STREAM = 3


@dataclass(frozen=True)
class SocketProtocol(EnumSerde[SocketProtocolValue]):
    @staticmethod
    def enum_type():
        return SocketProtocolValue


@dataclass
class StartServer(DeriveSerde):
    """
    Response: ServerOpened | OperationFailure
    """

    descriptor: MessageDescriptor
    protocol: SocketProtocol
    port: TunnelPort


@dataclass
class CloseServer(DeriveSerde):
    """
    Response: OperationSuccess | OperationFailure
    """

    descriptor: MessageDescriptor
    server: ServerDescriptor


@dataclass
class ConnectSocket(DeriveSerde):
    """
    Response: SocketConnected | OperationFailure
    """

    descriptor: MessageDescriptor
    protocol: SocketProtocol
    remote: NodeId
    port: TunnelPort


@dataclass
class SendData(DeriveSerde):
    """
    Response: OperationSuccess | OperationFailure
    """

    descriptor: MessageDescriptor
    socket: SocketDescriptor
    payload: RemainingBytes


@dataclass
class CloseSocket(DeriveSerde):
    """
    Response: OperationSuccess | OperationFailure
    """

    descriptor: MessageDescriptor
    socket: SocketDescriptor


@dataclass
class Terminate(DeriveSerde):
    pass


RequestBodyTypes = (
    SendHello
    | StartServer
    | CloseServer
    | ConnectSocket
    | CloseSocket
    | SendData
    | Terminate
)


@dataclass(frozen=True)
class RequestMessage(Variant[RequestBodyTypes]):
    @staticmethod
    def variant_types():
        return list(RequestBodyTypes.__args__)


@dataclass
class OperationSuccess(DeriveSerde):
    descriptor: MessageDescriptor


@dataclass(frozen=True)
class TraceStr(Utf8):
    @staticmethod
    def length_type():
        return UInt32


@dataclass
class OperationFailure(DeriveSerde):
    descriptor: MessageDescriptor
    trace: TraceStr


@dataclass
class ServerStarted(DeriveSerde):
    descriptor: MessageDescriptor
    server: ServerDescriptor


@dataclass
class ServerClosed(DeriveSerde):
    server: ServerDescriptor


@dataclass
class ServerSocketConnected(DeriveSerde):
    server: ServerDescriptor
    socket: SocketDescriptor
    protocol: SocketProtocol
    remote: NodeId
    port: TunnelPort


@dataclass
class SocketConnected(DeriveSerde):
    descriptor: MessageDescriptor
    socket: SocketDescriptor
    protocol: SocketProtocol
    remote: NodeId
    port: TunnelPort


@dataclass
class SocketReceived(DeriveSerde):
    socket: SocketDescriptor
    payload: RemainingBytes


@dataclass
class SocketClosed(DeriveSerde):
    socket: SocketDescriptor


ResponseBodyTypes = (
    OperationSuccess
    | OperationFailure
    | ServerStarted
    | ServerClosed
    | ServerSocketConnected
    | SocketConnected
    | SocketReceived
    | SocketClosed
)


@dataclass(frozen=True)
class ResponseMessage(Variant[ResponseBodyTypes]):
    @staticmethod
    def variant_types():
        return list(ResponseBodyTypes.__args__)


type ServerEvent = ServerSocketConnected | ServerClosed

type SocketEvent = SocketReceived | SocketClosed
